fem/convergence.cpp

```diff
@@ -0,0 +1,297 @@
+#include "convergence.hpp"
+
+using namespace std;
+
+
+namespace mfem
+{
+
+void ConvergenceStudy::Reset()
+{
+   counter=0;
+   dcounter=0;
+   fcounter=0;
+   cont_type=-1;
+   print_flag=1;
+   L2Errors.SetSize(0);
+   L2Rates.SetSize(0);
+   DErrors.SetSize(0);
+   DRates.SetSize(0);
+   EnErrors.SetSize(0);
+   EnRates.SetSize(0);
+   DGFaceErrors.SetSize(0);
+   DGFaceRates.SetSize(0);
+   ndofs.SetSize(0);
+}
+
+double ConvergenceStudy::GetNorm(GridFunction * gf, Coefficient * scalar_u,
+                                 VectorCoefficient * vector_u)
+{
+   bool norm_set = false;
+   double norm=0.0;
+   int order = gf->FESpace()->GetOrder(0);
+   int order_quad = std::max(2, 2*order+1);
+   const IntegrationRule *irs[Geometry::NumGeom];
+   for (int i=0; i < Geometry::NumGeom; ++i)
+   {
+      irs[i] = &(IntRules.Get(i, order_quad));
+   }
+#ifdef MFEM_USE_MPI
+   ParGridFunction * pgf = dynamic_cast<ParGridFunction *>(gf);
+   if (pgf)
+   {
+      ParMesh * pmesh = pgf->ParFESpace()->GetParMesh();
+      if (scalar_u)
+      {
+         norm = ComputeGlobalLpNorm(2.0,*scalar_u,*pmesh,irs);
+      }
+      else if (vector_u)
+      {
+         norm = ComputeGlobalLpNorm(2.0,*vector_u,*pmesh,irs);
+      }
+      norm_set = true;
+   }
+#endif
+   if (!norm_set)
+   {
+      Mesh * mesh = gf->FESpace()->GetMesh();
+      if (scalar_u)
+      {
+         norm = ComputeLpNorm(2.0,*scalar_u,*mesh,irs);
+      }
+      else if (vector_u)
+      {
+         norm = ComputeLpNorm(2.0,*vector_u,*mesh,irs);
+      }
+   }
+   return norm;
+}
+
+void ConvergenceStudy::AddL2Error(GridFunction * gf,
+                                  Coefficient * scalar_u, VectorCoefficient * vector_u)
+{
+   int tdofs=0;
+#ifdef MFEM_USE_MPI
+   ParGridFunction * pgf = dynamic_cast<ParGridFunction *>(gf);
+   if (pgf)
+   {
+      MPI_Comm comm = pgf->ParFESpace()->GetComm();
+      int rank;
+      MPI_Comm_rank(comm, &rank);
+      print_flag = 0;
+      if (rank==0) { print_flag = 1; }
+      tdofs = pgf->ParFESpace()->GlobalTrueVSize();
+   }
+#endif
+   if (!tdofs) { tdofs = gf->FESpace()->GetTrueVSize(); }
+   ndofs.Append(tdofs);
+   double L2Err;
+   if (scalar_u)
```

Comment:
Maybe I am in the minority here, but I think it is a lot more clear here and elsewhere to use polymorphism on the methods for the different kinds on coefficients:
GetNorm(GridFunction * gf, Coefficient * scalar_u)
GetNorm(GridFunction * gf, VectorCoefficient * vector_u)

When you have both of them on the call line it begs the question as to what the the method should do if you define both of them.  At the very least MFEM should error out if both of them are defined.

Reply:
Hi @acfisher, 
if both of them are set it will do only the Coefficient part (from the if/else if statement). I see your point though...I tried to do it this way to keep code replication minimal and since the method is private there is no case where both are set. I can always change it to use polymorphism or add a check if both are set...whatever you guys decide i am fine with it.

Socratis